skbio/stats/composition.py

```diff
@@ -2058,3 +2060,489 @@ def dirmult_ttest(
+    _covariate_list.remove("Intercept")
+
+    output = []
+    for response_var in table.columns:
+        # mixed effects code is obtained here:
+        # http://stackoverflow.com/a/22439820/1167475
+        stats_formula = "%s ~ %s" % (response_var, formula)
+        model = MixedLM.from_formula(
+            formula=stats_formula,
+            data=merged_data,
+            groups=groups,
+            re_formula=re_formula,
+            vc_formula=vc_formula,
+            **kwargs,
+        )
+
+        results = model.fit(reml=reml, method=method, **fit_kwargs)
+        summary = results.summary()
+
+        for var_name in _covariate_list:
+            try:
+                _lme_coef, _lme_25, _lme_975 = np.nan, np.nan, np.nan
+                _lme_coef = float(summary.tables[1]["Coef."][var_name])
+                _lme_25 = float(summary.tables[1]["[0.025"][var_name])
+                _lme_975 = float(summary.tables[1]["0.975]"][var_name])
+
+                individual_results = {
+                    FEATUREID: response_var,
+                    COVARIATE: var_name,
+                    LOG2FC: _lme_coef,
+                    CI25: _lme_25,
+                    CI975: _lme_975,
+                    PVALUE: results.pvalues[var_name],
+                }
+            except Exception as e:
+                if type(e) is not ValueError:
+                    print("Something went really wrong, details below: ")
+                    print(type(e))
+                    print(e)
+
+                    individual_results = {
+                        FEATUREID: response_var,
+                        COVARIATE: var_name,
+                        LOG2FC: np.nan,
+                        CI25: np.nan,
+                        CI975: np.nan,
+                        PVALUE: np.nan,
+                    }
+
+                else:
+                    measures = {
+                        LOG2FC: np.nan,
+                        CI25: np.nan,
+                        CI975: np.nan,
+                    }
+                    mapping = {
+                        LOG2FC: _lme_coef,
+                        CI25: _lme_25,
+                        CI975: _lme_975,
+                    }
+
+                    if len(summary.tables) >= 2:
+                        table = summary.tables[1]
+
+                        for key_final, key_table in mapping.items():
+                            if key_table in table and var_name in table[key_table]:
+                                measures[key_final] = table[key_table][var_name]
+
+                    individual_results = {
+                        "FeatureID": response_var,
+                        "Covariate": var_name,
+                        "pvalue": results.pvalues[var_name],
+                    }
+
+                    individual_results.update(measures)
+
+            output.append(individual_results)
+
+    return (output, submodels, _covariate_list)
+
+
+def _obtain_dir_table(data, pseudocount, rng):
```

Comment:
Can we do a refactor of the existing `dirmult_test` and replace all of the posterior calculations there?

https://github.com/scikit-bio/scikit-bio/blob/ffc47bf6af7d7948249e03e6b61d7848161323fc/skbio/stats/composition.py#L1993-L1996

https://github.com/scikit-bio/scikit-bio/blob/ffc47bf6af7d7948249e03e6b61d7848161323fc/skbio/stats/composition.py#L2006-L2011

It will add another sanity check to make sure that this code is working as expected. Plus it will help clean up the dirmult_ttest as well (there will be less redundancy)

Reply:
Yes, I have refactored `dirmult_ttest` to use the new function and removed the redundant code. It continues to work as expected.